File: tmdb-spring-boot-starter/src/test/java/dev/reuss/tmdb/spring/boot/TmdbClientAutoConfigurationTest.java
package dev.reuss.tmdb.spring.boot;

import dev.reuss.tmdb.TmdbClient;
import org.junit.jupiter.api.Test;
import org.springframework.boot.autoconfigure.AutoConfigurations;
import org.springframework.boot.test.context.runner.ApplicationContextRunner;

import static org.assertj.core.api.Assertions.assertThat;

class TmdbClientAutoConfigurationTest {

    private final ApplicationContextRunner contextRunner = new ApplicationContextRunner()
            .withConfiguration(AutoConfigurations.of(TmdbClientAutoConfiguration.class));

    @Test
    void createsTmdbClientWhenAccessTokenIsConfigured() {
        contextRunner
                .withPropertyValues("tmdb.access-token=test-token")
                .run(context -> assertThat(context).hasSingleBean(TmdbClient.class));
    }

    @Test
    void doesNotCreateTmdbClientWhenCustomBeanExists() {
        TmdbClient customClient = TmdbClient.builder()
                .accessToken("custom-token")
                .build();

        contextRunner
                .withBean(TmdbClient.class, () -> customClient)
                .withPropertyValues("tmdb.access-token=test-token")
                .run(context -> {
                    assertThat(context).hasSingleBean(TmdbClient.class);
                    assertThat(context.getBean(TmdbClient.class)).isSameAs(customClient);
                });
    }

    @Test
    void failsWhenAccessTokenIsMissing() {
        contextRunner.run(context -> assertThat(context).hasFailed());
    }

    @Test
    void bindsProperties() {
        contextRunner
                .withPropertyValues(
                        "tmdb.access-token=test-token",
                        "tmdb.default-language=de-DE",
                        "tmdb.default-region=DE",
                        "tmdb.connect-timeout=2s",
                        "tmdb.request-timeout=5s"
                )
                .run(context -> {
                    assertThat(context).hasSingleBean(TmdbClient.class);

                    TmdbProperties properties = context.getBean(TmdbProperties.class);
                    assertThat(properties.accessToken()).isEqualTo("test-token");
                    assertThat(properties.defaultLanguage()).isEqualTo("de-DE");
                    assertThat(properties.defaultRegion()).isEqualTo("DE");
                    assertThat(properties.connectTimeout()).hasSeconds(2);
                    assertThat(properties.requestTimeout()).hasSeconds(5);
                });
    }
}

File: tmdb-spring-boot-starter/src/test/kotlin/dev/reuss/tmdb/spring/boot/TmdbClientAutoConfigurationTest.kt
package dev.reuss.tmdb.spring.boot

import dev.reuss.tmdb.TmdbClient
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.boot.autoconfigure.AutoConfigurations
import org.springframework.boot.test.context.runner.ApplicationContextRunner

class TmdbClientAutoConfigurationTest {
    private val contextRunner =
        ApplicationContextRunner()
            .withConfiguration(
                AutoConfigurations.of(TmdbClientAutoConfiguration::class.java),
            )

    @Test
    fun createsTmdbClientWhenAccessTokenIsConfigured() {
        contextRunner
            .withPropertyValues("tmdb.access-token=test-token")
            .run { context ->
                assertThat(context)
                    .hasSingleBean(TmdbClient::class.java)
            }
    }

    @Test
    fun doesNotCreateTmdbClientWhenCustomBeanExists() {
        val customClient =
            TmdbClient
                .builder()
                .accessToken("custom-token")
                .build()

        contextRunner
            .withBean(
                TmdbClient::class.java,
                { customClient },
            ).withPropertyValues("tmdb.access-token=test-token")
            .run { context ->
                assertThat(context)
                    .hasSingleBean(TmdbClient::class.java)

                assertThat(
                    context.getBean(TmdbClient::class.java),
                ).isSameAs(customClient)
            }
    }

    @Test
    fun failsWhenAccessTokenIsMissing() {
        contextRunner.run { context ->
            assertThat(context).hasFailed()
        }
    }

    @Test
    fun bindsProperties() {
        contextRunner
            .withPropertyValues(
                "tmdb.access-token=test-token",
                "tmdb.default-language=de-DE",
                "tmdb.default-region=DE",
                "tmdb.connect-timeout=2s",
                "tmdb.request-timeout=5s",
            ).run { context ->
                assertThat(context)
                    .hasSingleBean(TmdbClient::class.java)

                val properties =
                    context.getBean(TmdbProperties::class.java)

                assertThat(properties.accessToken)
                    .isEqualTo("test-token")

                assertThat(properties.defaultLanguage)
                    .isEqualTo("de-DE")

                assertThat(properties.defaultRegion)
                    .isEqualTo("DE")

                assertThat(properties.connectTimeout)
                    .hasSeconds(2)

                assertThat(properties.requestTimeout)
                    .hasSeconds(5)
            }
    }
}
